zerver/views/registration.py

```diff
@@ -1121,21 +1134,72 @@ def realm_import_post_process(
     if request.method == "POST":
         form = ImportRealmOwnerSelectionForm(request.POST)
         if form.is_valid():
+            # This is a highly sensitive code path, since what we're
+            # about to do is take control of another user account AND
+            # promote that account to organization owner.
+            #
+            # We need this code path for Slack import if the importing
+            # user's account doesn't exist in the Slack import, but we
+            # must be VERY careful to make sure we're in that situation.
+
+            # Validate that this PreregistrationRealm object is in the
+            # expected state, with no user matching the email address,
+            # and not having been fully associated with the created realm.
+            #
+            # TODO: Could we have `created_realm` be already set, so
+            # we're not relying on the less precise linkage of `string_id`?
+            assert preregistration_realm.data_import_metadata["no_user_matching_email"]
+            assert preregistration_realm.created_realm is None
+            assert preregistration_realm.status != confirmation_settings.STATUS_USED
+
+            # This email is the imported account that the importing user has
+            # selected to become their account.
+            #
+            # TODO: Is there any reason not to switch this to user IDs?
             email = form.cleaned_data["email"]
-            user = get_user_by_delivery_email(email, realm)
-            user.delivery_email = preregistration_realm.email
-            user.role = UserProfile.ROLE_REALM_OWNER
-            user.save(update_fields=["delivery_email", "role"])
+
+            # Validate that a normal user account that can login was selected.
+            importing_user = get_user_by_delivery_email(email, realm)
```

Comment:
I renamed the local variables to `importing_user` across the various code paths, since it felt easier to read that way.